Bluetooth: controller: remove legacy LLCP code

kruithofa opened the pull request:
This PR removes the legacy LLCP controller code.
The code for delaying host notifications, which is only used for the legacy controller is also removed.
The babblesim tests are updated, and the tests for the legacy controller are removed.


Signed-off-by: Andries Kruithof <andries.kruithof@nordicsemi.no>

fixes https://github.com/zephyrproject-rtos/zephyr/issues/49573

Commit:
Bluetooth: controller: removing legacy LLCP code

This commit removes the legacy LLCP code including the
Kconfig option
It also updates the babblesim tests, and removes the
tests for the legacy controller

Signed-off-by: Andries Kruithof <andries.kruithof@nordicsemi.no>

Commit:
Blueotooth: controller: removed delayed notification code

The code that was added for delaying notifications to the host,
which was only implemented for legacy controller, is removed
in this commit

Signed-off-by: Andries Kruithof <andries.kruithof@nordicsemi.no>

Commit:
Bluetooth: controller: remove code related to advanced scheduling

This commit removes the code related to advanced scheduling from
ull_sched.c, since this code only works for the legacy controller.
Implementing advanced scheduling for the refactored LLCP is tracked
in its own issue

Signed-off-by: Andries Kruithof <andries.kruithof@nordicsemi.no>

Commit:
Bluetooth: controller: remove dependency for slot reservation

The slot reservation code is only valid for the refactored LLCP;
a guard was in place to ensure that it was not compiled in for
legacy controller code. This commit removes this guard

Signed-off-by: Andries Kruithof <andries.kruithof@nordicsemi.no>

Commit:
Bluetooth: controller: remove refs to LLCP_LEGACY after rebase

Some recent PRs included conditional compilation for
KCONFIG_BT_LL_SW_LLCP_LEGACY, which must be removed.

Signed-off-by: Andries Kruithof <andries.kruithof@nordicsemi.no>

thoh-ot commented on subsys/bluetooth/controller/ll_sw/ull_conn.c:
```diff
@@ -8495,10 +2473,15 @@ void ull_dle_local_tx_update(struct ll_conn *conn, uint16_t tx_octets, uint16_t
 
 void ull_dle_init(struct ll_conn *conn, uint8_t phy)
 {
+	/*
+	 * TODO:
+	 * legacy code uses the maximum of the time for 1M phy and actual phy
+	 * for max_time_min, to be verified if that is required here as well
+	 */
```
?

thoh-ot reviewed (changes requested):
Could not add marker to unchanged lines, so i'll add my findings here:
subsys/bluetooth/controller/ll_sw/ull_conn.c:
  140-145: Obsolete defines
  147-150: Remove comment

kruithofa commented on subsys/bluetooth/controller/ll_sw/ull_conn.c:
```diff
@@ -2637,5357 +1832,233 @@ static void conn_cleanup_iso_cis_released_cb(struct ll_conn *conn)
+			    (void **)&tx);
+	while (link) {
+		uint8_t idx;
+		struct lll_tx *lll_tx2;
```
Because one of the checks in CI complains if the variable has the same name as the struct name. Agreed that the suffix 2 does not give a better name though

kruithofa commented on subsys/bluetooth/controller/ll_sw/ull_central.c:
```diff
@@ -424,24 +342,20 @@ uint8_t ll_create_connection(uint16_t scan_interval, uint16_t scan_window,
+	conn_lll->dle.eff.max_tx_time = MAX(conn_lll->dle.eff.max_tx_time,
+					    PDU_DC_MAX_US(PDU_DC_PAYLOAD_SIZE_MIN,
+							  lll->phy));
+	conn_lll->dle.eff.max_rx_time = MAX(conn_lll->dle.eff.max_rx_time,
+					    PDU_DC_MAX_US(PDU_DC_PAYLOAD_SIZE_MIN,
+							  lll->phy));
 #endif /* CONFIG_BT_CTLR_ADV_EXT */
-	max_tx_time = conn_lll->max_tx_time;
-	max_rx_time = conn_lll->max_rx_time;
-#else /* !CONFIG_BT_CTLR_PHY */
-	max_tx_time = PDU_DC_MAX_US(PDU_DC_PAYLOAD_SIZE_MIN, PHY_1M);
-	max_rx_time = PDU_DC_MAX_US(PDU_DC_PAYLOAD_SIZE_MIN, PHY_1M);
-#endif /* !CONFIG_BT_CTLR_PHY */
-#else /* !CONFIG_BT_CTLR_DATA_LENGTH */
+	max_tx_time = conn_lll->dle.eff.max_tx_time;
+	max_rx_time = conn_lll->dle.eff.max_rx_time;
+#else /* CONFIG_BT_CTLR_DATA_LENGTH */
```
I think it is the diff-tool that makes it look incorrect

thoh-ot replied:
I think i was referring to the fact that somehow you have added a comment with my github username in it (line 345):
`/* TODO(thoh-ot): Not entirely sure this is correct */`

Commit:
Bluetooth: controller: update permutation script

Updated the script that compiles all different permutations so
that it starts with the correct configuration file

Signed-off-by: Andries Kruithof <andries.kruithof@nordicsemi.no>

thoh-ot commented on subsys/bluetooth/controller/ll_sw/ull_sched.c:
```diff
@@ -243,21 +226,10 @@ void ull_sched_mfy_win_offset_use(void *param)
-	 * TODO: update when updating the connection update procedure
+	 * TODO: update calculationg of the win_offset
+	 * when updating the connection update procedure
+	 * see the legacy code from Zephyr v3.2 for inspiration
```
Should this refer to v3.3?

kruithofa replied:
correct. 3.3 is the latest release with legacy in it

thoh-ot commented on subsys/bluetooth/controller/ll_sw/ull_sched.c:
```diff
@@ -266,370 +238,36 @@ void ull_sched_mfy_free_win_offset_calc(void *param)
 
 	/*
 	 * TODO: update when updating the connection update procedure
+	 * see the legacy code from Zephyr v3.2 for inspiration
```
Should this refer to v3.3?

thoh-ot commented on subsys/bluetooth/controller/ll_sw/ull_sched.c:
```diff
@@ -266,370 +238,36 @@ void ull_sched_mfy_free_win_offset_calc(void *param)
-	 * TODO: update when updating the connection update procedure
+	 * TODO: update calculation of win_offset when
+	 * updating the connection update procedure
+	 * see the legacy code from Zephyr v3.2 for inspiration
```
Should this refer to v3.3?

thoh-ot commented on subsys/bluetooth/controller/ll_sw/ull_sched.c:
```diff
@@ -266,370 +238,36 @@ void ull_sched_mfy_free_win_offset_calc(void *param)
-#endif /* CONFIG_BT_LL_SW_LLCP_LEGACY */
+/*
+ * TODO: probably we need a function for calculating the window offset
+ * see the legacy code from Zephyr v3.2 for inspiration
```
Should this refer to v3.3?

thoh-ot commented on subsys/bluetooth/controller/ll_sw/ull_conn.c:
```diff
@@ -193,60 +125,28 @@ static uint8_t force_md_cnt_calc(struct lll_conn *lll_conn, uint32_t tx_rate);
 	uint8_t pool[sizeof(memq_link_t) *
 		     (CONN_DATA_BUFFERS +
-		      CONN_TX_CTRL_BUFFERS)];
+			     LLCP_TX_CTRL_BUF_COUNT)];
```
Indentation seems wrong

kruithofa replied:
fixed

kruithofa commented on subsys/bluetooth/controller/ll_sw/ull_central.c:
```diff
@@ -426,24 +337,19 @@ uint8_t ll_create_connection(uint16_t scan_interval, uint16_t scan_window,
 #endif
 
-#if defined(CONFIG_BT_LL_SW_LLCP_LEGACY)
+	/* TODO: Not entirely sure this is correct */
```
comment removed

kruithofa commented:
rebased

kruithofa commented:
rebased

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_conn_iso.c:
```diff
@@ -294,35 +292,24 @@ uint8_t ll_conn_iso_accept_timeout_set(uint16_t timeout)
+	node_rx->hdr.handle = cis->lll.acl_handle;
 
-	/* TODO: Send CIS_ESTABLISHED with status != 0 in error scenarios */
-	node_rx->hdr.handle = 0xFFFF;
-	node_rx->hdr.rx_ftr.param = cis;
-
-	est = (void *)node_rx->pdu;
-	est->status = 0;
-	est->cis_handle = cis_lll->handle;
-
-	ll_rx_put_sched(node_rx->hdr.link, node_rx);
-#endif /* !CONFIG_BT_LL_SW_LLCP_LEGACY */
+	ull_rx_put_sched(node_rx->hdr.link, node_rx);
```
Something is not right here, this is not there for LLCP in the LHS

kruithofa replied:
removed

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_conn.c:
```diff
@@ -193,60 +125,28 @@ static uint8_t force_md_cnt_calc(struct lll_conn *lll_conn, uint32_t tx_rate);
-#define CONN_ENC_REQ_BUFFERS CONFIG_BT_CTLR_LLCP_CONN
-#else
 #define CONN_ENC_REQ_BUFFERS 0
-#endif
 #define CONN_DATA_BUFFERS (CONFIG_BT_BUF_ACL_TX_COUNT + CONN_ENC_REQ_BUFFERS)
```
Remove CONN_ENC_REQ_BUFFERS

kruithofa replied:
removed

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_conn.c:
```diff
@@ -193,60 +125,28 @@ static uint8_t force_md_cnt_calc(struct lll_conn *lll_conn, uint32_t tx_rate);
-#else /* !CONFIG_BT_LL_SW_LLCP_LEGACY */
-#define CONN_TX_CTRL_BUFFERS LLCP_TX_CTRL_BUF_COUNT
-#endif /* !CONFIG_BT_LL_SW_LLCP_LEGACY */
 #define CONN_TX_CTRL_BUF_SIZE MROUND(offsetof(struct node_tx, pdu) + \
 				     offsetof(struct pdu_data, llctrl) + \
 				     PDU_DC_CTRL_TX_SIZE_MAX)
```
Is this used?

kruithofa replied:
not anymore. removed

cvinayak reviewed (approved):
Changes LGTM. I have rebased all my current work in progress with the changes in this PR, tested iso_broadcast and Central multiple CIS babblesim tests that I am working on; no visible regressions.

erbr-ot commented on tests/bluetooth/bsim/_compile_permutate_kconfigs.sh:
```diff
@@ -62,8 +62,8 @@ perm_compile() {
-    local app=tests/bluetooth/bsim/ll/edtt/hci_test_app
-    local conf_file=prj_dut.conf
+    local app=tests/bluetooth/bsim/edtt_ble_test_app/hci_test_app
+    local conf_file=prj_dut_llcp.conf
```
Hmm, still carrying the 'llcp' in the naming - shouldn't we loose that? Perhaps do this in a subsequent PR, as this was introduced to distinguish between legacy and refactored

kruithofa replied:
I'll do it in a subsequent PR

erbr-ot reviewed (approved):
With the comment re. _llcp in the naming re. edtt tests/scripts/conf I'm good with this.
